classes/curlmanager_security_helper.php

```diff
@@ -22,150 +29,134 @@
+        global $DB;
 
-            $plugin = $this->getcomponentbycodepath($rootcodepath);
-            if ($plugin === false) {
-                $plugin = '';
-            }
+        $shoudlog = get_config('tool_curlmanager', 'loggingenabled');
+        if (!$shoudlog) {
+            return;
+        }
 
-            // Suggest to deduplicate on host, plugin and codepath.
-            // Check if the host, plugin and codepath exists already.
-            // Add a new record if not exist.
-            // Otherwise update the reocrd with count+1 and timeupdated field.
-            $record = $DB->get_records('tool_curlmanager',
-                ['host' => $host, 'plugin' => $plugin],
-                '',
-                'id, count'
-            );
-
-            if (count($record) > 0) {
-                $record = current($record);
-                $data = new \stdClass();
-                $data->id = $record->id;
-                $data->count = $record->count + 1;
-                $data->codepath = $formattedbacktrace;
-                $data->urlallowed = $returnvalue['allowed'] ? 1 : 0;
-                $data->urlblocked = $urlblocked ? 1 : 0;
-                $data->timeupdated = time();
-                $DB->update_record('tool_curlmanager', $data);
-
-            } else {
-                $data = new \stdClass();
-                $data->plugin = $plugin;
-                $data->codepath = $formattedbacktrace;
-                $data->url = $urlstring;
-                $data->host = $host;
-                $data->urlallowed = $returnvalue['allowed'] ? 1 : 0;
-                $data->urlblocked = $urlblocked ? 1 : 0;
-                $data->count = 1;
-                $data->timecreated = time();
-                $data->timeupdated = time();
-                $DB->insert_record('tool_curlmanager', $data);
-            }
+        // Build stacktrace.
+        $rootcodepath = '';
+        $trace = debug_backtrace();
+        $formattedbacktrace = format_backtrace(debug_backtrace(), true);
+        $lasttrace = count($trace) - 1;
+        if (isset($trace[$lasttrace]['file'])) {
+            $rootcodepath = $trace[$lasttrace]['file'];
         }
 
-        // If allow host is enabled and the host is not in the allowed host list, return true.
-        if ($returnvalue['allowhostenabled'] && $returnvalue['allowed'] === false) {
-            return true;
+        // Parse plugin.
+        $plugin = $this->getcomponentbycodepath($rootcodepath);
+        if ($plugin === false) {
+            $plugin = '';
         }
 
-        return false;
+        // Get path (without any query params).
+        $urlstring = self::sanitise_url($url);
+        $reference = self::get_reference($url, $formattedbacktrace);
+
+        // Upsert log using reference to de-dup.
+        $record = $DB->get_record('tool_curlmanager', ['reference' => $reference], 'id,count', IGNORE_MISSING);
+
+        if (!empty($record)) {
+            $DB->update_record('tool_curlmanager', [
```

Comment:
So I was a little suspect of this previously overwriting data that wasnt required. I think we should remove urlallowed and urlblocked from the upserts, and instead add it to the reference hash so that if it changes, it will be logged to a different line, but that way we don't need to overwrite it on every curl invocation

thoughts?

Reply:
agreed - i'll add it in